### PR TITLE
Stability: Fix Recent Test Flakes

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Fix a number of recent flakes that we have seen in CI. This started as a fix purely for the consul_test, our most painful flake. But I encountered a few others along the way and decided to try to resolve them

# Context

### Why Will This Fix the Flake?
I don't know (haven't investigated) whether this will fix the flake. What I did was convert the existing tests to use the TestContext framework. The hope is that that framework has been proven to be better about applying resources in order to avoid flakes.

I plan on leaving this PR open for a few days to see how consistently it passes CI. If we see consistent passes, we can merge this. If there is no improvement in the frequency, I will investigate this further.

_Update_: Over the course of this PR, CI has run >20 times (for new commits) and while I hit other flakes (that were resolved in this PR) I did not see the consul_test flake once, which makes me hopeful that this change will resolve the flake.

### Consul Run Options
We previously defined certain consul properties in our test runner (consulWatch, consulDnsAddress). I updated this to use the pattern we're trying to follow, where the Settings object holds the source of. truth, and we build our options from those settings.

### TestContextWithConsul
I followed a similar pattern that I had in https://github.com/solo-io/gloo/pull/8021. Specifically, we have only 1 set of tests which interface with a consul service, so it didn't make sense to bake it into the core TestContext. I built a simple extension which includes the ConsulService and a method to run the service.

### Testing
You can run these tests locally. I focused the entire file and then ran:
```
ENVOY_IMAGE_TAG=1.14.0-beta8 TEST_PKG=./test/e2e RUN_CONSUL_TESTS=1 INVALID_TEST_REQS=skip make test
```

`INVALID_TEST_REQS=skip` ensures that the test which requires linux will be auto-skipped

## ContainSubstrings matcher panic
When building this PR, I encountered https://github.com/solo-io/gloo/issues/8003#issuecomment-1491073016, which was a panic that occurred from our `ContainSubstrings` custom matcher. That wrapper is just syntactic sugar around the built-in `ContainSubstring`, where we And all of the matchers together. The panic occurred when we try to and a list of 1 item. The solution was to just handle the case separately

## Cleanup Test Services
As part of this recent work, we have been cleaning up (commenting, fixing, removing unnecessary configuration) from our Vault. Consul and Envoy test services. We have a number of other tests services in that package, that aren't used (and haven't been used for years). 

Below are the list of files I removed, and a brief rationale for why I felt safe removing it:
- [functions.go](https://github.com/solo-io/gloo/blob/master/test/services/functions.go): This file was entirely commented out, and the last modification was from 2018
- [mesh.go](https://github.com/solo-io/gloo/blob/master/test/services/mesh.go): This file is built to test a service mesh use case, which Gloo Edge does not solve anymore. This file was unused, except by our Prometheus test service, which I removed.
- [prometheus.go](https://github.com/solo-io/gloo/blob/master/test/services/prometheus.go):  This was not used anywhere and the last edit was 2018.
- [gloo.go](https://github.com/solo-io/gloo/blob/master/test/services/gloo.go):  This file was entirely commented out.

## Zipkin Test Flake
I encountered a test flake in the Zipkin test, and created https://github.com/solo-io/gloo/issues/8043 to track it. I have a small fix that I believe should resolve this.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/7932
resolves https://github.com/solo-io/gloo/issues/8043

### CODE REVIEW

File: test/services/gateway.go [hunk 1]
```diff
@@ -199,16 +201,14 @@ type What struct {
...
 type RunOptions struct {
-	NsToWrite        string
-	NsToWatch        []string
-	WhatToRun        What
-	GlooPort         int32
-	ValidationPort   int32
-	RestXdsPort      int32
-	Settings         *gloov1.Settings
-	KubeClient       kubernetes.Interface
-	ConsulClient     consul.ConsulWatcher
```
Comment (sam-heilbron) [inline test/services/gateway.go:210]: ConsulClient and ConsulDnsAddress were deleted, explained further in PR context

File: changelog/v1.14.0-beta20/consul-e2e-flake.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+changelog:
+  - type: NON_USER_FACING
+    issueLink: https://github.com/solo-io/gloo/issues/7932
+    resolvesIssue: true
+    description: Fix a flake in our consul_test which would cause inconsistent behavior in CI.
+  - type: NON_USER_FACING
+    issueLink: https://github.com/solo-io/gloo/issues/8003
+    resolvesIssue: false
+    description: >- 
+      Avoid a panic in the `ContainSubstrings` Gomega Matcher, by handling the case of a single
```
Comment (gunnar-solo) [inline changelog/v1.14.0-beta20/consul-e2e-flake.yaml:10]: _optimally_ I'd have this be its own PR.  If no-one else is unhappy with it here, though:  great!

File: projects/gloo/cli/pkg/cmd/create/create_test.go [hunk 3]
```diff
@@ -1,37 +1,41 @@
...
 import (
+	"context"
 	"log"
-	"os"
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 	"github.com/solo-io/gloo/projects/gloo/cli/pkg/helpers"
 	"github.com/solo-io/gloo/projects/gloo/cli/pkg/testutils"
+	testutils2 "github.com/solo-io/gloo/test/testutils"
```
Comment (gunnar-solo) [inline projects/gloo/cli/pkg/cmd/create/create_test.go:11]: it'd make me _feel_ better if this import pair weren't `testutils` and `testutils2`.

`gloo_cli_testutils` and `gloo_testutils`, perhaps?

File: projects/gloo/pkg/bootstrap/utils_test.go [hunk 4]
```diff
@@ -146,24 +145,27 @@ var _ = Describe("Utils", func() {
...
 
+		var (
+			ctx    context.Context
+			cancel context.CancelFunc
+		)
+
 		BeforeEach(func() {
-			if os.Getenv("RUN_CONSUL_TESTS") != "1" {
+
+			if !testutils.IsEnvTruthy(testutils.RunConsulTests) {
```
Comment (gunnar-solo) [inline projects/gloo/pkg/bootstrap/utils_test.go:155]: I appreciate this change!  Thank you for cleaning up our scattered hardcoded strings

**Review by gunnar-solo — COMMENTED** [review]
some small, scattered thoughts

File: projects/gloo/cli/pkg/cmd/create/create_test.go (hunk 3 above)
Comment (sam-heilbron) [reply projects/gloo/cli/pkg/cmd/create/create_test.go:11]: Yeah good call! I saw this and thought about it and then was lazy. Thanks for keeping me honest...I will push an update

File: changelog/v1.14.0-beta20/consul-e2e-flake.yaml (hunk 2 above)
Comment (sam-heilbron) [reply changelog/v1.14.0-beta20/consul-e2e-flake.yaml:10]: Yeah I like that point. My fear was just that without this fix in the PR, I might hit that flake/failure and it would slow down development. Happy to move it if you have strong feelings

File: test/e2e/consul_test.go [hunk 5]
```diff
@@ -3,297 +3,208 @@ package e2e_test
...
 	"time"
 
+	"github.com/golang/protobuf/ptypes/duration"
+	v1 "github.com/solo-io/gloo/projects/gateway/pkg/api/v1"
+	"github.com/solo-io/gloo/test/e2e"
+	"github.com/solo-io/gloo/test/gomega/matchers"
+
 	"github.com/solo-io/gloo/test/testutils"
 
 	consul2 "github.com/solo-io/gloo/projects/gloo/pkg/plugins/consul"
```
Comment (ianmacclancy) [inline test/e2e/consul_test.go:15]: this follows the same pattern that Gunnar [commented on](https://github.com/solo-io/gloo/pull/8032/files#r1156410160) - if changing there we could update this to gloo_consul_utils. Name up to you

File: projects/gloo/cli/pkg/cmd/create/create_test.go (hunk 3 above)
Comment (sam-heilbron) [reply projects/gloo/cli/pkg/cmd/create/create_test.go]: I've always struggled with the name convention of package aliases in Go. I came across https://github.com/kubernetes/kubernetes/pull/35258 and figured we tend to follow this, so I used the following alias:
- `gloo_cli_testutils` -> `cliutils`
- `gloo_testutils` -> `testutils`

I didn't want to make this change too broadly as part of this PR, but I think it would be nice to adopt a convention informally within the team and then through some requirements in CI. 

How do you feel about us taking the following steps:
1. We reach out to the team to see what they feel our convention should be
2. I update https://github.com/solo-io/gloo/issues/6686 to reflect this

File: test/e2e/consul_test.go (hunk 5 above)
Comment (sam-heilbron) [reply test/e2e/consul_test.go]: Yeah good call! I updated this to `consulplugin` and `consulapi` to try to identify the source of the package. What do you think?

File: projects/gloo/cli/pkg/cmd/create/create_test.go (hunk 3 above)
Comment (gunnar-solo) [reply projects/gloo/cli/pkg/cmd/create/create_test.go]: I'm perfectly happy with the "not fixing things broadly"--just _this_ is good.

----
> 1. We reach out to the team to see...

this feels like it'd be a good topic for a lightning knowledge share.  Ex: 5 minute talk/demo as Monday "demo".  Re:  types of imports, why "." is bad practice, go standards, how imports in other files in same package function...

If that's too much:  "eh"

----
> I update ...

perfect!

**Review by gunnar-solo — APPROVED** [review]
